### PR TITLE
fix asan check in c++ gradients

### PR DESCRIPTION
**Pull request by vnghia**
@saxenasaurabh 

Let's see if this PR fixes the problems with asan. There are some more leaks indeed, but I couldn't find a fix for them and they seem related to other components of the code. Here is the list:
 
- `tensorflow::BuildImmediateExecutionContext`
- `tensorflow::gradients::Tape::ComputeGradient`
- `tensorflow::EagerContext::CreateOperation`
- `tensorflow::gradients::TapeOperation::Execute`

### CODE REVIEW

File: tensorflow/c/eager/BUILD
Comment (vnvo2409) [inline tensorflow/c/eager/BUILD:417]: Here

File: tensorflow/c/experimental/gradients/BUILD
Comment (vnvo2409) [inline tensorflow/c/experimental/gradients/BUILD:181]: Here

**Review by vnghia — COMMENTED** [review]
Could you please elaborate what does this `args` mean ? Maybe it is related to the leaks above and I could do something. Thank you !

File: tensorflow/c/eager/BUILD
Comment (saxenasaurabh) [inline tensorflow/c/eager/BUILD:419]: Let's leave the `no_cuda_asan` in the BUILD targets for now unless you have been able to test that the leaks are gone. I can try to remove this and test locally once your changes are merged. It might take longer to merge this if the tests are still failing since we will have to wait for the internal tests, add this back, re-pull etc.

**Comment by saxenasaurabh** [comment]
Thanks for finding and fixing this. Would you mind describing / documenting how you are debugging mem leaks. It may be useful for future contributors.

> Could you please elaborate what does this args mean ? Maybe it is related to the leaks above and I could do something. Thank you !

This basically says do not perform heap leak checking for the entire program but only for annotated code blocks (using an internal API). Since there are no annotated code-blocks in this test this basically skips checking for leaks. We can remove this once the leaks are gone.

File: tensorflow/c/eager/BUILD
Comment (vnvo2409) [reply tensorflow/c/eager/BUILD]: Done

**Comment by vnvo2409** [comment]
> Thanks for finding and fixing this. Would you mind describing / documenting how you are debugging mem leaks. It may be useful for future contributors.

I am not sure where the documenting should go. I think it is best if we could add this [option](https://stackoverflow.com/a/57733619) to our `.bazelrc` file.

**Comment by saxenasaurabh** [comment]
> > Thanks for finding and fixing this. Would you mind describing / documenting how you are debugging mem leaks. It may be useful for future contributors.
> 
> I am not sure where the documenting should go. I think it is best if we could add this [option](https://stackoverflow.com/a/57733619) to our `.bazelrc` file.

cc: @mihaimaruseac

File: tensorflow/c/eager/gradient_checker.cc
Comment (saxenasaurabh) [inline tensorflow/c/eager/gradient_checker.cc:86]: Could this just be `absl::MakeSpan(outputs)` and we can drop the next line?

**Comment by vnvo2409** [comment]
> This basically says do not perform heap leak checking for the entire program but only for annotated code blocks (using an internal API). Since there are no annotated code-blocks in this test this basically skips checking for leaks. We can remove this once the leaks are gone.

I found this site https://gperftools.github.io/gperftools/heap_checker.html . It seems related to that `args`. Could you check it out ?

**Comment by saxenasaurabh** [comment]
> > This basically says do not perform heap leak checking for the entire program but only for annotated code blocks (using an internal API). Since there are no annotated code-blocks in this test this basically skips checking for leaks. We can remove this once the leaks are gone.
> 
> I found this site https://gperftools.github.io/gperftools/heap_checker.html . It seems related to that `args`. Could you check it out ?

Aha, yep looks like the correct link.

**Review by saxenasaurabh — APPROVED** [review]

**Comment by vnvo2409** [comment]
FYI,

When building asan without debugging, e.g only `--copt -fsanitize=address --linkopt -fsanitize=address`, `custom_gradient_test` pass. However, when adding `--copt -O1 --copt -g -c dbg`, there is a leak in `tensorflow::BuildImmediateExecutionContext` ( it is probably why we need `heap-check=local` ).

My guess is we could drop `no_cuda_asan` for `gradient_checker_test`.

`nn_grad_test` has some leaks even without debugging symbol, so we will have to leave it as-is.

**Pull request merged**
